packages/react-client/src/ReactFlightClient.js

```diff
@@ -854,18 +900,90 @@ export function processBinaryChunk(
-      readFinalStringChunk(stringDecoder, chunk.subarray(0, linebreak));
-    processFullRow(response, fullrow);
-    response._partialRow = '';
-    chunk = chunk.subarray(linebreak + 1);
-    linebreak = chunk.indexOf(10); // newline
+  let i = 0;
+  while (i < chunk.length) {
+    let lastIdx = -1;
+    switch (response._rowState) {
+      case ROW_ID: {
+        const byte = chunk[i++];
+        if (byte === 58 /* ":" */) {
+          // Finished the rowID, next we'll parse the tag.
+          response._rowState = ROW_TAG;
+        } else {
+          response._rowID =
+            (response._rowID << 4) | (byte > 96 ? byte - 87 : byte - 48);
+        }
+        continue;
+      }
+      case ROW_TAG: {
+        const resolvedRowTag = chunk[i];
+        if (resolvedRowTag === 84 /* "T" */) {
+          response._rowTag = resolvedRowTag;
+          response._rowState = ROW_LENGTH;
+          i++;
+        } else if (resolvedRowTag > 64 && resolvedRowTag < 91 /* "A"-"Z" */) {
+          response._rowTag = resolvedRowTag;
+          response._rowState = ROW_CHUNK_BY_NEWLINE;
+          i++;
+        } else {
+          response._rowTag = 0;
+          response._rowState = ROW_CHUNK_BY_NEWLINE;
+          // This was an unknown tag so it was probably part of the data.
+        }
+        continue;
+      }
+      case ROW_LENGTH: {
+        const byte = chunk[i++];
+        if (byte === 44 /* "," */) {
+          // Finished the rowLength, next we'll buffer up to that length.
+          response._rowState = ROW_CHUNK_BY_LENGTH;
+        } else {
+          response._rowLength =
+            (response._rowLength << 4) | (byte > 96 ? byte - 87 : byte - 48);
+        }
+        continue;
+      }
+      case ROW_CHUNK_BY_NEWLINE: {
+        // We're looking for a newline
+        lastIdx = chunk.indexOf(10 /* "\n" */, i);
+        break;
+      }
+      case ROW_CHUNK_BY_LENGTH: {
+        // We're looking for the remaining byte length
+        const rowLength = response._rowLength;
+        if (i + rowLength <= chunk.length) {
+          lastIdx = i + rowLength;
+        }
+        break;
+      }
+    }
+    if (lastIdx > -1) {
+      // We found the last chunk of the row
+      const lastChunk = chunk.slice(i, lastIdx);
+      processFullRow(
+        response,
+        response._rowID,
+        response._rowTag,
+        response._buffer,
+        lastChunk,
+      );
+      // Reset state machine for a new row
+      response._rowState = ROW_ID;
+      response._rowTag = 0;
+      response._rowID = 0;
+      response._rowLength = 0;
+      response._buffer.length = 0;
+      i = lastIdx + 1;
+    } else {
+      // The rest of this row is in a future chunk. We stash the rest of the
+      // current chunk until we can process the full row.
+      const remainingSlice = chunk.slice(i);
+      response._buffer.push(remainingSlice);
+      // Update how many bytes we're still waiting for.
+      response._rowLength -= remainingSlice.length;
```

Comment:
Maybe it isn't worth the extra condition but seems this should be just for the chunks by length case. Might be worth a comment that you are eliding the conditional since the subtraction is faster and the property is ignored in row by new line cases?